common/interpolation/README.md

```diff
@@ -0,0 +1,109 @@
+Methods except for tridiagonal matrix algorithm exists in `spline_interpolation` package, which has been removed from Autoware.
+
+| Method                            | Calculation time |
+| --------------------------------- | ---------------- |
+| Tridiagonal Matrix Algorithm      | 0.007 [ms]       |
+| Preconditioned Conjugate Gradient | 0.024 [ms]       |
+| Successive Over-Relaxation        | 0.074 [ms]       |
+
+### Spline Interpolation Algorithm
+
+Assuming that the size of `base_keys` ($x_i$) and `base_values` ($y_i$) are $N + 1$, we aim to calculate spline interpolation with the following equation to interpolate between $y_i$ and $y_{i+1}$.
+
+$$
+Y_i(x) = a_i (x - x_i)^3 + b_i (x - x_i)^2 + c_i (x - x_i) + d_i \ \ \ (i = 0, \dots, N-1)
+$$
+
+Constraints on spline interpolation are as follows.
+The number of constraints is $4N$, which is equal to the number of variables of spline interpolation.
+
+$$
+\begin{align}
+Y_i (x_i) & = y_i \ \ \ (i = 0, \dots, N-1) \\
+Y_i (x_{i+1}) & = y_{i+1} \ \ \ (i = 0, \dots, N-1) \\
+Y'_i (x_{i+1}) & = Y'_{i+1} (x_{i+1}) \ \ \ (i = 0, \dots, N-2) \\
+Y''_i (x_{i+1}) & = Y''_{i+1} (x_{i+1}) \ \ \ (i = 0, \dots, N-2) \\
+Y''_0 (x_0) & = 0 \\
+Y''_{N-1} (x_N) & = 0
+\end{align}
+$$
+
+According to [this article](https://www.mk-mode.com/rails/docs/INTERPOLATION_SPLINE.pdf), spline interpolation is formulated as the following linear equation.
+
+$$
+\begin{align}
+ \begin{pmatrix}
+    2(h_0 + h_1) & h_1 \\
+    h_0 & 2 (h_1 + h_2) & h_2 & & O \\
+        &     &     & \ddots \\
+    O &     &     &       & h_{N-2} & 2 (h_{N-2} + h_{N-1})
+ \end{pmatrix}
+ \begin{pmatrix}
+    v_1 \\ v_2 \\ v_3 \\ \vdots \\ v_{N-1}
+ \end{pmatrix}=
+ \begin{pmatrix}
+    w_1 \\ w_2 \\ w_3 \\ \vdots \\ w_{N-1}
+ \end{pmatrix}
+\end{align}
+$$
+
+where
+
+$$
+\begin{align}
+h_i & = x_{i+1} - x_i \ \ \ (i = 0, \dots, N-1) \\
+w_i & = 6 \left(\frac{y_{i+1} - y_{i+1}}{h_i} - \frac{y_i - y_{i-1}}{h_{i-1}}\right) \ \ \ (i = 1, \dots, N-1)
+\end{align}
+$$
+
+The coefficient matrix of this linear equation is tridiagonal matrix. Therefore, it can be solve with tridiagonal matrix algorithm, which can solve linear equations without gradient descent methods.
+
+Solving this linear equation with tridiagonal matrix algorithm, we can calculate coefficients of spline interpolation as follows.
+
+$$
+\begin{align}
+a_i & = \frac{v_{i+1} - v_i}{6 (x_{i+1} - x_i)} \ \ \ (i = 0, \dots, N-1) \\
+b_i & = \frac{v_i}{2} \ \ \ (i = 0, \dots, N-1) \\
+c_i & = \frac{y_{i+1} - y_i}{x_{i+1} - x_i} - \frac{1}{6}(x_{i+1} - x_i)(2 v_i + v_{i+1}) \ \ \ (i = 0, \dots, N-1) \\
+d_i & = y_i \ \ \ (i = 0, \dots, N-1)
+\end{align}
+$$
+
+### Tridiagonal Matrix Algorithm
+
+We solve tridiagonal linear equation according to [this article](https://mathlang.hatenablog.com/entry/2018/10/14/232741) where variables of linear equation are expressed as follows in the implementation.
```

Comment:
This is the URL of Japanese article.
Is it OK?
@kenji-miyake

Reply:
@tkimura4 @kenji-miyake 
As long as this package referred to the articles, we should write regardless of language, I think.

Reply:
https://github.com/autowarefoundation/autoware.universe/issues/88